### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by luciotato**

### CODE REVIEW

File: fun-token-contract/.cargo/config [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[build]
+rustflags = ["-C", "link-args=-s"]
```
Comment (robert-zaremba) [inline fun-token-contract/.cargo/config:2]: let's put the whole thing in `/devx` directory. This contract is not part of the principal project, it's used for testing and simulation.

File: fun-token-contract/build.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+sh.ls('./target/wasm32-unknown-unknown/release/*.wasm').map(src => {
+  const output = path.basename(src)
+    .replace('.wasm', '-rs.wasm')
+    .replace(/_/g, '-')
+
+  console.log(`\ncopying [ ${src} ] to [ out/${output} ]`);
+
+  sh.cp(src, `${outdir}/${output}`)
+})
+*/
```
Comment (robert-zaremba) [inline fun-token-contract/build.js:27]: this file is not needed. Let's use simple shell script or a Makefile.

File: fun-token-contract/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+        if refund_amount > 0 {
+            env::log(format!("Refunding {} tokens for storage", refund_amount).as_bytes());
+            Promise::new(env::predecessor_account_id()).transfer(refund_amount);
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(test)]
+mod tests {
```
Comment (robert-zaremba) [inline fun-token-contract/src/lib.rs:293]: This is a copy, paste - we don't need tests here. Shall we remove it?

File: fun-token-contract/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+    pub fn get_allowance(&self, owner_id: AccountId, escrow_account_id: AccountId) -> U128 {
+        assert!(
+            env::is_valid_account_id(escrow_account_id.as_bytes()),
+            "Escrow account ID is invalid"
+        );
+        self.get_account(&owner_id)
+            .get_allowance(&escrow_account_id)
+            .into()
+    }
+}
```
Comment (robert-zaremba) [inline fun-token-contract/src/lib.rs:242]: we need one more function here:

```
pub fn mint(&mut self)
```
Which will mint 10 tokens to `env::predecessor_account_id()` and increase the `total_balance`.

**Review by robert-zaremba — CHANGES_REQUESTED** [review]

File: contract/.vscode/tasks.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	"version": "2.0.0",
+	"tasks": [
+		{
+			"type": "shell",
+			"command":"cargo",
+			"args": [
+				"+nightly",
+                "test",
+                // "--",
+				// "arg1"
```
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/.vscode/tasks.json:11]: to add more params later

File: contract/Cargo.toml [hunk 6]
```diff
@@ -10,10 +10,20 @@ crate-type = ["cdylib", "rlib"]
 [dependencies]
 serde = { version = "*", features = ["derive"] }
 serde_json = "*"
-# borsh = "^0.7.0"
+##borsh = "^0.7.0"
```
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/Cargo.toml:13]: I included a lot of dependencies in order to do simulation-tests -- that means it was commented originally

File: contract/build.sh [hunk 7]
```diff
@@ -1 +1 @@
-env 'RUSTFLAGS=-C link-arg=-s' cargo build --target wasm32-unknown-unknown --release
+env 'RUSTFLAGS=-C link-arg=-s' cargo +stable build --target wasm32-unknown-unknown --release
```
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/build.sh:1]: yes. https://github.com/near-examples/simulation-testing#gotchas

File: contract/src/lib.rs [hunk 8]
```diff
@@ -29,11 +29,11 @@ const SINGLE_CALL_GAS: u64 = 200_000_000_000_000;
...
 #[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
 pub struct PoolInfo {
-    near_bal: Balance,
-    token_bal: Balance,
+    pub near_bal: Balance,
+    pub token_bal: Balance,
     /// total amount of participation shares. Shares are represented using the same amount of
     /// tailing decimals as the NEAR token, which is 24
-    total_shares: Balance,
+    pub total_shares: Balance,
```
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/src/lib.rs:36]: I need that to make asserts in the tests. That's inside the contract (WASM), it's not really "public" as in outside access.  It's fine.

File: contract/src/lib.rs [hunk 9]
```diff
@@ -944,7 +945,9 @@ mod tests {
         token1.inc_allowance(t.clone(), token_deposit.into());
 
         ctx.set_vmc_deposit(near_deposit);
-        c.add_liquidity(t.clone(), ctx.token_supply, ctx.token_supply);
+        let max_token_deposit = token_deposit;
+        let min_shares_required = near_deposit;
```
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/src/lib.rs:949]: I think those assignments will get optimized away by LLVM. I'll check that and if it's not the case I'll remove them.
The consts are there to help reading the code.

File: contract/.vscode/tasks.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply contract/.vscode/tasks.json:11]: I would prefer to not commit `.vscode` into the repository. If there is anything useful we can create scripts (eg in bash).

File: contract/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply contract/Cargo.toml:13]: I know, but this change doesn't bring anything nor it's a style improvement.

File: contract/.vscode/tasks.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/.vscode/tasks.json:11]: those are 2 files with a few bytes each, and are very useful to build and test in vscode.
We have bigger fish to fry Robert.
Please gloss over this small stuff and let's focus on the core functions.

File: contract/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (luciotato) [reply contract/Cargo.toml:13]: Robert, the core code functions are not working.
please don't waste my and your time over a #

File: fun-token-contract/.cargo/config (hunk 1 above)
Comment (luciotato) [reply fun-token-contract/.cargo/config:2]: It's really complex as it is. Let's leave it here until the core CLP functions pass all the integration tests

File: fun-token-contract/build.js (hunk 2 above)
Comment (luciotato) [reply fun-token-contract/build.js:27]: small stuff. leave it there. I'll clean it up *after* demo day

**Review by robert-zaremba — APPROVED** [review]
@luciotato , I did small cleaning, approving and merging. The reason I did cleaning is to keep a correct trace of history. Otherwise it's getting messy if we want to browse changes. Also, I want to keep remote code to minimum - it will create less confusion. So, if we copy the code form other repo, then copy the minimum and adding ref link.

Thanks for making this PR.

**Pull request merged**